File: src/java1/algorithms/graph/CourseSchedule.java
package java1.algorithms.graph;

import java.util.*;

public class CourseSchedule {
    //BFS:==> TC: O(V+E) SC: O(V)
    private static boolean canFinish(int numCourses, int[][] prerequisites){
        Map<Integer, List<Integer>> adj = new HashMap<>(); 
        int[] indegree = new int[numCourses];

        for(int i=0; i< numCourses; i++) {
            adj.put(i, new ArrayList<Integer>());
        }

        for(int i=0; i< prerequisites.length; i++) {
            adj.get(prerequisites[i][1]).add(prerequisites[i][0]);
            indegree[prerequisites[i][0]]++;
        }

        int enrolled = 0;
        Queue<Integer> queue= new LinkedList<Integer>();
        for(int node=0; node< indegree.length; node++) {
            if(indegree[node] == 0) {
                queue.add(node);
                enrolled++;
            }
        }

        while(!queue.isEmpty()) {
            int currCourse = queue.poll();
            for(int neighbour: adj.get(currCourse)) {
                indegree[neighbour]--;
                if(indegree[neighbour] == 0) {
                    queue.add(neighbour);
                    enrolled++;
                }
            }
        }
        return enrolled == numCourses;
    }

    public static void main(String[] args) {
        int numCourses1 =6;
        int[][] prerequisites1 = {{0,4}, {0,5}, {1, 3}, {1, 4}, {2, 5}, {3, 2}};
        int numCourses2 =2;
        int[][] prerequisites2 = {{0,1}, {1,0}};
        System.out.println(canFinish(numCourses1, prerequisites1));
        System.out.println(canFinish(numCourses2, prerequisites2));
    }
    
}

File: src/javascript/algorithms/graph/courseSchedule.js
 //BFS:==> TC: O(V+E) SC: O(V)(Use Kahn's algorithm to see if topological order is possible or not)
function canFinish(numCourses, prerequisites) {
    let adjMap = Array(numCourses).fill(0).map(()=> []);
    let inDegree = Array(numCourses).fill(0);

    for(let [to, from] of prerequisites) {
        adjMap[from].push(to);
        inDegree[to]++;
    }

    let queue = [];
    let enrolled = 0;
    for(let i=0; i<numCourses;i++) {
        if(inDegree[i] === 0) {
            queue.push(i);
            enrolled++;
        }
    }
    
    while(queue.length >0) {
        let currentCourse = queue.pop();
        for(const neighbour of adjMap[currentCourse]) {
            inDegree[neighbour]--;
            if(inDegree[neighbour] === 0) {
                queue.push(neighbour);
                enrolled++;
            }
        }
    }
    return enrolled === numCourses;
}

const numCourses1 =6;
const prerequisites1 = [[0, 4], [0, 5], [1, 3], [1, 4], [2, 5], [3, 2]];
const numCourses2 =2;
const prerequisites2 = [[0, 1], [1, 0]];
console.log(canFinish(numCourses1, prerequisites1));
console.log(canFinish(numCourses2, prerequisites2));